contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,660 @@
+        where
+            Self: AccessControlled,
+        {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::Owner(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            self.access_control = access_control;
+            Ok(())
+        }
+
+        /// Returns current address of the AccessControl contract that holds the account priviledges for this DEX
+        #[ink(message)]
+        pub fn access_control(&self) -> AccountId {
+            self.access_control
+        }
+
+        /// Terminates the contract.
+        ///
+        /// Can only be called by the contract's Owner.
+        #[ink(message)]
+        pub fn terminate(&mut self) -> Result<(), DexError> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::Owner(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            self.env().terminate_contract(caller)
+        }
+
+        /// Returns own code hash
+        #[ink(message)]
+        pub fn code_hash(&self) -> Result<Hash, DexError> {
+            self.env()
+                .own_code_hash()
+                .map_err(|why| DexError::InkEnv(format!("Can't retrieve own code hash: {:?}", why)))
+        }
+
+        fn new_init(&mut self, ubik: AccountId, cyberiad: AccountId, lono: AccountId) {
+            self.access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+            self.ubik = ubik;
+            self.cyberiad = cyberiad;
+            self.lono = lono;
+        }
+
+        /// Transfers a given amount of a PSP22 token to a specified using the callers own balance
+        fn transfer_tx(
+            &self,
+            token: AccountId,
+            to: AccountId,
+            amount: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_SELECTOR))
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
+
+        /// Transfers a given amount of a PSP22 token on behalf of a specified account to another account
+        ///
+        /// Will revert if not enough allowance was given to the caller prior to executing this tx
+        fn transfer_from_tx(
```

Comment:
okay, now I see your point from the other PR that we'd better extract these common cross-contract calls (mostly to PSP22)